IntegrationTests/UI.IntegrationTests/UserControls/CommitInfo/CommitInfoTests.cs

```diff
@@ -72,90 +54,131 @@ public void OneTimeTearDown()
+
+                var refs = commitInfo.GetTestAccessor().GetSortedTags();
+
+                refs.Count.Should().Be(5);
+                refs.Should().BeEquivalentTo(expected);
+            });
         }
 
         [Test]
         public void GetSortedTags_should_remove_duplicate_refs()
         {
-            _gitExecutable.StageOutput("for-each-ref --sort=-taggerdate --format=\"%(refname)\" refs/tags/",
-                "refs/remotes/origin/master\nrefs/remotes/foo/duplicate\nrefs/remotes/foo/bar\nrefs/remotes/foo/duplicate\nrefs/remotes/foo/last"); // exact duplicates
-
-            var expected = new Dictionary<string, int>
+            RunCommitInfoTest(commitInfo =>
             {
-                ["refs/remotes/origin/master"] = 0,
-                ["refs/remotes/foo/duplicate"] = 1,
-                ["refs/remotes/foo/bar"] = 2,
-                ["refs/remotes/foo/last"] = 3,
-            };
-
-            var refs = _commitInfo.GetTestAccessor().GetSortedTags();
+                _gitExecutable.StageOutput("for-each-ref --sort=-taggerdate --format=\"%(refname)\" refs/tags/",
+                    "refs/remotes/origin/master\nrefs/remotes/foo/duplicate\nrefs/remotes/foo/bar\nrefs/remotes/foo/duplicate\nrefs/remotes/foo/last"); // exact duplicates
+
+                var expected = new Dictionary<string, int>
+                {
+                    ["refs/remotes/origin/master"] = 0,
+                    ["refs/remotes/foo/duplicate"] = 1,
+                    ["refs/remotes/foo/bar"] = 2,
+                    ["refs/remotes/foo/last"] = 3,
+                };
+
+                var refs = commitInfo.GetTestAccessor().GetSortedTags();
+
+                refs.Count.Should().Be(4);
+                refs.Should().BeEquivalentTo(expected);
+            });
+        }
 
-            refs.Count.Should().Be(4);
-            refs.Should().BeEquivalentTo(expected);
+        private void RunCommitInfoTest(Action<GitUI.CommitInfo.CommitInfo> runTest)
+        {
+            UITest.RunControl(
+                createControl: form =>
+                {
+                    var uiCommandsSource = Substitute.For<IGitUICommandsSource>();
+                    uiCommandsSource.UICommands.Returns(x => _commands);
+
+                    // the following assignment of CommitInfo.UICommandsSource will already call this command
+                    _gitExecutable.StageOutput("for-each-ref --sort=-taggerdate --format=\"%(refname)\" refs/tags/", "");
+
+                    return new GitUI.CommitInfo.CommitInfo
+                    {
+                        Parent = form,
+                        UICommandsSource = uiCommandsSource
+                    };
+                },
+                runTestAsync: async commitInfo =>
+                {
+                    // Wait for pending operations so the Contol is loaded completely before testing it
+                    await AsyncTestHelper.JoinPendingOperationsAsync(AsyncTestHelper.UnexpectedTimeout);
```

Comment:
❔ Is this required for any of the tests? It looks like `GetSortedTags` produces a result synchronously and on-demand.

We should remove this if the tests don't need it.

Reply:
➡️ Another call to `GetSortedTags` on a background thread could invalidate the staged output here. Joining pending operations ensures a 1:1 mapping of staged output to results.

Reply:
as tried to explain in the comments:
```
        public CommitInfo()
        {
            InitializeComponent();
            InitializeComplete();

            UICommandsSourceSet += delegate
            {
                this.InvokeAsync(() =>
                {
                    UICommandsSource.UICommandsChanged += delegate { RefreshSortedTags(); };

                    // call this event handler also now (necessary for "Contained in branches/tags")
                    RefreshSortedTags();
                }).FileAndForget();
            };
```
plus
```
                    // the following assignment of CommitInfo.UICommandsSource will already call this command
                    _gitExecutable.StageOutput("for-each-ref --sort=-taggerdate --format=\"%(refname)\" refs/tags/", "");

                    return new GitUI.CommitInfo.CommitInfo
                    {
                        Parent = form,
                        UICommandsSource = uiCommandsSource
                    };
```